### DISCUSSION TITLE
Is it possible to disable a certain form of caching in RDF4J?

### DISCUSSION THREAD

**Discussion by DennisDosso** [eclipse-rdf4j/rdf4j#3621] [Q&A] [answered]
Hello,

I am a researcher at the University of Padua and I am testing the performances of certain databases such as RDF4J for a paper of mine. 
What I noticed is that, for example, when I execute one query on an RDF database (BSBM to be precise),  the first time the query requires circa 1s. The second time I execute the same query, the execution time goes down to 300ms, and it stays at that value for all successive executions. 

To be more precise, I am running the Java class that executes the query from my terminal. The class only executes the query once, then the connection to the database is closed. The same class, executed two times in a row, returns two different results (1s and 300ms, as I was saying).

Another interesting thing that I noticed is that if I run the same class after a certain amount of time, it goes back to being executed in 1s (no other operation was done in the terminal during that time). It appears as if RDF4J is saving some information on disk or somewhere else (whatever that may be) that is deleted after a certain amount of time.

Since I would like to avoid having in my measurements the effect generated by forms of caching, is there a way to disable this option, whatever this is? I hope I explained myself here. 
Thanks for all the help, 
Best.

**Comment by hmottestad** [comment, marked as answer]
Great that you are benchmarking RDF4J!

RDF4J doesn't cache the query results between queries (except for the ElasticsearchStore).

The NativeStore has a number of caches for reducing IO during read and write operations. Some of these are configurable while others are not. Here are some that are configurable: 
- https://rdf4j.org/javadoc/latest/
- https://rdf4j.org/javadoc/latest/

Your operating system also has a disk cache, which will affect your performance. Your CPU also has a number of caches that will affect your performance too. 

Do keep in mind that the Java Virtual Machine uses just-in-time compilation, so you need to warm up the system before you can make any measurements. Take a look at the OpenJDK Java Microbenchmarking Harness: https://github.com/openjdk/jmh

Lastly I would like to point out that caching is a fundamental aspect to any DB. Disabling caching will make your results very unrealistic and not reflect the performance that a user experiences.

**Reply by DennisDosso** [reply]
Thank you very much for the answer. It seems it will give me a good direction and help me perform experiments that make sense. 
Have a great day.

**Reply by hmottestad** [reply]
Sorry about the links, the javadocs uses iframes :(

Here are the correct links:
 - https://rdf4j.org/javadoc/latest/org/eclipse/rdf4j/sail/nativerdf/config/NativeStoreConfig.html
 - https://rdf4j.org/javadoc/latest/org/eclipse/rdf4j/sail/nativerdf/config/NativeStoreSchema.html

You might also be interested in our general documentation: https://rdf4j.org/documentation/

**Reply by abrokenjester** [reply]
I'd also suggest running each query more than once: do a few warmup runs, but then also run it a couple of times more for measurements, and take the average of those runs as your benchmark result.
